interfaces/tests/file_int/file_int.c

```diff
@@ -0,0 +1,427 @@
+    wake_by_address_single(&(read_context->value));
+}
+
+
+BEGIN_TEST_SUITE(file_int)
+
+
+TEST_SUITE_INITIALIZE(a)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+}
+
+TEST_SUITE_CLEANUP(b)
+{
+
+    TEST_MUTEX_DESTROY(g_testByTest);
+}
+
+TEST_FUNCTION_INITIALIZE(c)
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+}
+
+TEST_FUNCTION_CLEANUP(d)
+{
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/*Tests_SRS_FILE_43_003: [If a file with name full_file_name does not exist, file_create shall create a file with that name.]*/
+/*Tests_SRS_FILE_43_001 : [file_create shall open the file named full_file_name for asynchronous operationsand return its handle.]*/
+/*Tests_SRS_FILE_43_006 : [file_destroy shall wait for all pending I / O operations to complete.]*/
+/*Tests_SRS_FILE_43_007 : [file_destroy shall close the file handle handle.]*/
+/*Tests_SRS_FILE_43_014 : [file_write_async shall enqueue a write request to write source's content to the position offset in the file. ]*/
+/*Tests_SRS_FILE_43_041:[If position + size is greater than the size of the file and the call to write is succesfull, file_write_async shall grow the file to accomodate the write.]*/
+/*Tests_SRS_FILE_43_008 : [file_write_async shall call user_call_back passing user_context and success depending on the success of the asynchronous write operation.]*/
+/*Tests_SRS_FILE_43_030 : [file_write_async shall succeed and return FILE_WRITE_ASYNC_OK.]*/
+/*Tests_SRS_FILE_43_021 : [file_read_async shall enqueue a read request to read handle's content at position offset and write it to destination. ]*/
+/*Tests_SRS_FILE_43_016:[file_read_async shall call user_callback passing user_context and success depending on the success of the asynchronous read operation.]*/
+/*Tests_SRS_FILE_43_031 : [file_read_async shall succeed and return FILE_READ_ASYNC_OK.]*/
+TEST_FUNCTION(write_to_a_file_and_read_from_it)
+{
+    ///arrange
+    const int size = 5;
+
+    unsigned char source[] = "abcd";
+    WRITE_COMPLETE_CONTEXT write_context;
+    write_context.pre_callback_value = 41;
+    interlocked_exchange(&write_context.value, write_context.pre_callback_value);
+    write_context.post_callback_value = 42;
+
+    unsigned char destination[5];
+    READ_COMPLETE_CONTEXT read_context;
+    read_context.pre_callback_value = 43;
+    interlocked_exchange(&read_context.value, read_context.pre_callback_value);
+    read_context.post_callback_value = 44;
+
+
+
+    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(NULL);
+    FILE_HANDLE file_handle = file_create(execution_engine, "write_to_a_file_and_read_from_it.txt", NULL, NULL);
+
+    ///act
+    FILE_WRITE_ASYNC_RESULT write_result = file_write_async(file_handle, source, size, 0, write_callback, &write_context);
+    wait_on_address(&write_context.value, &write_context.pre_callback_value, UINT32_MAX);

```

Comment:
>wait_on_address [](start = 4, length = 15)

Not sufficient due to spurious wakeups. Needs to be:

```
do
{
    wait_on_address
    interlocked_add to check the value
} while (not expected)
```

Ideally you would use InterlockedHL, but that needs to be updated still.

So I would probably put a helper in this file that does that (wait for value to be equal). Then we could remove the helper and switch to InterlockedHL some time later.

Reply:
done

---
In reply to: [456714198](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456714198) [](ancestors = 456714198)